File: src/types/hospital.ts
// src/types/hospital.ts

// ============================================
// Core hospital (admin / dashboard)
// ============================================

export interface Hospital {
  id: string;
  name: string;
  registrationNumber: string;
  type: 'government' | 'private' | 'charitable' | 'general' | 'multispecialty' | 'community' | 'teaching' | 'specialized';

  phone: string;
  emergencyPhone: string;
  email: string;
  website?: string;

  address: HospitalAddress;

  rating: number;
  reviewCount: number;
  totalBeds: number;
  occupiedBeds: number;
  availableBeds: number;

  icuTotalBeds: number;
  icuOccupiedBeds: number;
  icuAvailableBeds: number;
  icuWithVentilator: number;
  icuWithoutVentilator: number;

  emergencyServiceStatus: 'active' | 'busy' | 'unavailable';
  emergencyResponseTime: string;
  ambulanceCount: number;
  ambulanceAvailable: number;

  oxygenCylinders: OxygenStock;
  bloodBank: BloodBank;
  departments: Department[];

  totalDoctors: number;
  totalNurses: number;
  totalStaff: number;

  services: string[];
  facilities: string[];
  insuranceAccepted: string[];

  workingHours: WorkingHours;
  visitingHours: VisitingHours;

  coordinates: {
    latitude: number;
    longitude: number;
  };

  isActive: boolean;
  isVerified: boolean;
  createdAt: string;
  updatedAt: string;

  // Extended properties for hospital service
  location?: HospitalLocation & { coordinates?: { lat: number; lng: number } };
  rating?: number;
  reviewsCount?: number;
  beds?: {
    total: number;
    available: number;
    icu: { total: number; available: number };
    emergency?: { total: number; available: number };
  };
  emergency?: boolean;
  verified?: boolean;
  distance?: number;
  eta?: number;
  ambulanceAvailable?: boolean;
  oxygenAvailable?: boolean;
  pharmacy?: boolean;
  departments?: Department[] | string[];
}

export interface HospitalAddress {
  street: string;
  city: string;
  state: string;
  zipCode: string;
  country: string;
  landmark?: string;
}

export interface OxygenStock {
  total: number;
  available: number;
  inUse: number;
  reserved: number;
  lastRefilled: string;
  supplier: string;
  cylinderTypes: CylinderType[];
}

export interface CylinderType {
  type: string;
  capacity: string;
  total: number;
  available: number;
}

export interface BloodBank {
  isAvailable: boolean;
  bloodStock: BloodStock[];
  lastUpdated: string;
  totalUnits: number;
  expiryAlerts: number;
}

export interface BloodStock {
  bloodGroup: BloodGroup;
  units: number;
  expiryDate: string;
  status: 'sufficient' | 'low' | 'critical' | 'out-of-stock';
}

export type BloodGroup = 'A+' | 'A-' | 'B+' | 'B-' | 'AB+' | 'AB-' | 'O+' | 'O-';

export interface Department {
  id: string;
  name: string;
  description: string;
  headDoctor: string;
  headDoctorId: string;
  totalBeds: number;
  availableBeds: number;
  totalDoctors: number;
  totalNurses: number;
  services: string[];
  isActive: boolean;
  timings: string;
}

export interface WorkingHours {
  monday: DayHours;
  tuesday: DayHours;
  wednesday: DayHours;
  thursday: DayHours;
  friday: DayHours;
  saturday: DayHours;
  sunday: DayHours;
}

export interface DayHours {
  open: string;
  close: string;
  isOpen: boolean;
}

export interface VisitingHours {
  morning: VisitingTimeSlot;
  evening: VisitingTimeSlot;
}

export interface VisitingTimeSlot {
  start: string;
  end: string;
}

export interface HospitalDoctor {
  id: string;
  hospitalId: string;
  name: string;
  email: string;
  phone: string;
  specialization: string;
  department: string;
  designation: string;
  qualification: string;
  experience: number;
  consultationFee: number;
  availability: DoctorAvailability[];
  rating: number;
  reviewCount: number;
  status: 'active' | 'on-leave' | 'inactive';
  joinedDate: string;
  profileImage?: string;
}

export interface DoctorAvailability {
  day: string;
  startTime: string;
  endTime: string;
  maxPatients: number;
}

export interface BloodDonor {
  id: string;
  name: string;
  email: string;
  phone: string;
  bloodGroup: BloodGroup;
  age: number;
  gender: string;
  address: string;
  lastDonationDate?: string;
  totalDonations: number;
  isVerified: boolean;
  isAvailable: boolean;
  medicalConditions: string[];
  registrationDate: string;
  status: 'pending' | 'approved' | 'rejected' | 'blocked';
}

export interface BloodRequest {
  id: string;
  requestNumber: string;
  patientName: string;
  patientAge: number;
  bloodGroup: BloodGroup;
  units: number;
  hospitalName: string;
  doctorName: string;
  urgency: 'normal' | 'urgent' | 'emergency';
  status: 'pending' | 'approved' | 'processing' | 'fulfilled' | 'rejected';
  requestDate: string;
  requiredDate: string;
  reason: string;
  approvedBy?: string;
  fulfilledBy?: string;
}

export interface Ambulance {
  id: string;
  vehicleNumber: string;
  type: 'basic' | 'advanced' | 'cardiac' | 'neonatal' | 'mobile-icu';
  status: 'available' | 'on-call' | 'maintenance' | 'offline';
  currentLocation?: {
    latitude: number;
    longitude: number;
    address: string;
  };
  driverName: string;
  driverPhone: string;
  paramedicName?: string;
  equipment: string[];
  lastService: string;
  nextService: string;
  isACAvailable: boolean;
  hasOxygenSupport: boolean;
}

export interface AmbulanceRequest {
  id: string;
  patientName: string;
  patientPhone: string;
  pickupLocation: string;
  dropLocation: string;
  emergencyType: string;
  status: 'pending' | 'dispatched' | 'picked-up' | 'arrived' | 'cancelled';
  assignedAmbulance?: string;
  requestTime: string;
  dispatchTime?: string;
  estimatedArrival: string;
}

export interface EmergencyAnnouncement {
  id: string;
  title: string;
  message: string;
  type: 'general' | 'emergency' | 'blood-camp' | 'health-camp' | 'awareness';
  priority: 'low' | 'medium' | 'high' | 'critical';
  targetAudience: 'all' | 'doctors' | 'patients' | 'staff' | 'public';
  createdBy: string;
  createdAt: string;
  expiresAt?: string;
  isActive: boolean;
  attachments?: string[];
}

export interface ICUBed {
  id: string;
  bedNumber: string;
  hospitalId: string;
  hospitalName: string;
  type: 'general-icu' | 'cardiac-icu' | 'neuro-icu' | 'pediatric-icu' | 'neonatal-icu';
  status: 'available' | 'occupied' | 'reserved' | 'maintenance';
  patientName?: string;
  patientId?: string;
  admissionDate?: string;
  hasVentilator: boolean;
  hasMonitor: boolean;
  nurseStation: string;
  floor: string;
  dailyCharge: number;
}

export interface HospitalDashboardData {
  hospital: Hospital;
  todayStats: TodayStats;
  bedStats: BedStats;
  icuStats: ICUStats;
  bloodStats: BloodStats;
  oxygenStats: OxygenStats;
  emergencyStats: EmergencyStats;
  ambulanceStats: AmbulanceStats;
  recentActivities: HospitalActivity[];
  pendingApprovals: PendingApproval[];
  alerts: Alert[];
  analytics: HospitalAnalytics;
}

export interface TodayStats {
  totalPatients: number;
  newAdmissions: number;
  discharges: number;
  emergencies: number;
  surgeries: number;
  deaths: number;
  births: number;
}

export interface BedStats {
  total: number;
  occupied: number;
  available: number;
  reserved: number;
  occupancyRate: number;
  byDepartment: DepartmentBedStat[];
}

export interface DepartmentBedStat {
  department: string;
  total: number;
  occupied: number;
  available: number;
}

export interface ICUStats {
  total: number;
  occupied: number;
  available: number;
  onVentilator: number;
  criticalPatients: number;
  averageStay: string;
}

export interface BloodStats {
  totalUnits: number;
  expiringSoon: number;
  criticalGroups: BloodGroup[];
  todayRequests: number;
  todayDonations: number;
}

export interface OxygenStats {
  totalCylinders: number;
  inUse: number;
  available: number;
  reserved: number;
  daysLeft: number;
}

export interface EmergencyStats {
  todayEmergencies: number;
  activeEmergencies: number;
  averageResponseTime: string;
  ambulancesDispatched: number;
}

export interface AmbulanceStats {
  total: number;
  available: number;
  onCall: number;
  inMaintenance: number;
}

export interface HospitalActivity {
  id: string;
  type: 'admission' | 'discharge' | 'emergency' | 'surgery' | 'blood' | 'ambulance' | 'other';
  description: string;
  time: string;
  department?: string;
  user?: string;
}

export interface PendingApproval {
  id: string;
  type: 'doctor' | 'blood-donor' | 'blood-request' | 'leave';
  name: string;
  details: string;
  requestDate: string;
  priority: 'normal' | 'urgent';
}

export interface Alert {
  id: string;
  type: 'critical' | 'warning' | 'info' | 'success';
  message: string;
  time: string;
  isRead: boolean;
}

export interface HospitalAnalytics {
  patientFlow: {
    labels: string[];
    admissions: number[];
    discharges: number[];
  };
  bedOccupancy: {
    labels: string[];
    rates: number[];
  };
  bloodUsage: {
    labels: string[];
    used: number[];
    donated: number[];
  };
  revenue: {
    labels: string[];
    amount: number[];
  };
}

// ============================================
// Legacy listing / card types (public directory)
// ============================================

export type HospitalType =
  | 'general'
  | 'multispecialty'
  | 'teaching'
  | 'community'
  | 'specialty'
  | 'clinic'
  | 'trauma';

export interface BedInfo {
  type: string;
  total: number;
  occupied: number;
  available: number;
  pricePerDay: number;
  features?: string[];
}

export interface BedSummary {
  total: number;
  available: number;
  icu: { total: number; available: number };
  emergency: number;
}

export interface HospitalContact {
  phone: string;
  emergencyPhone?: string;
  email: string;
  website?: string;
}

export interface HospitalLocation {
  address: string;
  city: string;
  state: string;
  coordinates?: { lat: number; lng: number };
}

/** Hospital card / search listing shape used by mock data and public pages */
export interface HospitalListItem {
  id: string;
  name: string;
  type: HospitalType;
  location?: HospitalLocation;
  address?: {
    city?: string;
    state?: string;
    address?: string;
  };
  contact?: HospitalContact;
  beds?: BedInfo[] | BedSummary;
  services?: string[];
  specialties?: string[];
  doctorsCount?: number;
  rating: number;
  reviewCount: number;
  emergency?: boolean;
  emergencyAvailable?: boolean;
  ambulance?: boolean;
  ambulanceAvailable?: boolean;
  bloodBank?: boolean;
  bloodBankAvailable?: boolean;
  oxygen?: boolean;
  oxygenAvailable?: boolean;
  verified?: boolean;
  isVerified?: boolean;
  premium?: boolean;
  isPremium?: boolean;
  distance?: number;
  eta?: number;
  image?: string;
  images?: string[];
  established?: number;
  establishedYear?: number;
  accreditation?: string[];
}

export interface HospitalCardProps {
  hospital: HospitalListItem;
  onViewBeds?: (hospital: HospitalListItem) => void;
  variant?: 'default' | 'compact' | 'featured';
}

export interface BedBooking {
  id: string;
  hospitalId: string;
  bedType: string;
  patientId: string;
  patientName: string;
  admittedAt: string;
  expectedDischarge?: string;
  status: 'pending' | 'confirmed' | 'admitted' | 'discharged' | 'cancelled';
  totalCost?: number;
}

export interface BookBedPayload {
  hospitalId: string;
  bedType: string;
  patientId: string;
  patientName: string;
  admittedAt: string;
  expectedDischarge?: string;
}

export interface HospitalState {
  hospitals: HospitalListItem[];
  nearbyHospitals: HospitalListItem[];
  emergencyHospitals: HospitalListItem[];
  selectedHospital: HospitalListItem | null;
  beds: BedInfo[];
  isLoading: boolean;
  error: string | null;
}

File: models/hospital.py
from sqlalchemy import (
    Column, String, Integer, Boolean, DECIMAL, JSON, Text, Date, Time,
    Enum as SQLEnum, ForeignKey, Index
)
from sqlalchemy.dialects.mysql import CHAR
from ..core.database import Base
from .base import BaseModel


# ============================================
# HOSPITALS
# ============================================
class Hospital(Base, BaseModel):
    __tablename__ = "hospitals"

    admin_user_id = Column(CHAR(36), ForeignKey("users.id"), nullable=False, index=True)
    name = Column(String(255), nullable=False)
    registration_number = Column(String(100), unique=True, nullable=False)
    type = Column(SQLEnum("government", "private", "charitable", "multispecialty", "community", "teaching", "specialized", "general"))
    phone = Column(String(20), nullable=False)
    emergency_phone = Column(String(20))
    email = Column(String(255))
    website = Column(String(500))
    street = Column(String(255))
    city = Column(String(100), nullable=False, index=True)
    state = Column(String(100), nullable=False, index=True)
    zip_code = Column(String(20))
    country = Column(String(100), default="USA")
    latitude = Column(DECIMAL(10, 7))
    longitude = Column(DECIMAL(10, 7))
    total_beds = Column(Integer, default=0)
    available_beds = Column(Integer, default=0)
    icu_total_beds = Column(Integer, default=0)
    icu_available_beds = Column(Integer, default=0)
    icu_with_ventilator = Column(Integer, default=0)
    icu_without_ventilator = Column(Integer, default=0)
    ambulance_count = Column(Integer, default=0)
    ambulance_available = Column(Integer, default=0)
    emergency_service = Column(SQLEnum("active", "busy", "unavailable"), default="active")
    emergency_response_time = Column(String(20), default="15 min")
    total_doctors = Column(Integer, default=0)
    total_nurses = Column(Integer, default=0)
    total_staff = Column(Integer, default=0)
    rating = Column(DECIMAL(2, 1), default=0.0)
    review_count = Column(Integer, default=0)
    is_verified = Column(Boolean, default=False)
    is_active = Column(Boolean, default=True)
    verified_by = Column(CHAR(36), ForeignKey("users.id"))
    services = Column(JSON)
    facilities = Column(JSON)
    insurance_accepted = Column(JSON)
    working_hours = Column(JSON)
    visiting_hours = Column(JSON)


# ============================================
# HOSPITAL DEPARTMENTS
# ============================================
class HospitalDepartment(Base, BaseModel):
    __tablename__ = "hospital_departments"

    hospital_id = Column(CHAR(36), ForeignKey("hospitals.id", ondelete="CASCADE"), nullable=False, index=True)
    name = Column(String(150), nullable=False)
    description = Column(Text)
    head_doctor_id = Column(CHAR(36), ForeignKey("doctor_profiles.id"), nullable=True)
    total_beds = Column(Integer, default=0)
    available_beds = Column(Integer, default=0)
    total_doctors = Column(Integer, default=0)
    total_nurses = Column(Integer, default=0)
    services = Column(JSON)
    timings = Column(String(100))
    is_active = Column(Boolean, default=True)


# ============================================
# HOSPITAL DOCTORS (Affiliation)
# ============================================
class HospitalDoctor(Base, BaseModel):
    __tablename__ = "hospital_doctors"

    hospital_id = Column(CHAR(36), ForeignKey("hospitals.id", ondelete="CASCADE"), nullable=False, index=True)
    doctor_id = Column(CHAR(36), ForeignKey("doctor_profiles.id", ondelete="CASCADE"), nullable=False, index=True)
    department_id = Column(CHAR(36), ForeignKey("hospital_departments.id"), nullable=True)
    designation = Column(String(100))
    joining_date = Column(Date)
    is_active = Column(Boolean, default=True)

    __table_args__ = (Index("ix_hospital_doctor_unique", "hospital_id", "doctor_id", unique=True),)


# ============================================
# HOSPITAL BEDS
# ============================================
class HospitalBed(Base, BaseModel):
    __tablename__ = "hospital_beds"

    hospital_id = Column(CHAR(36), ForeignKey("hospitals.id", ondelete="CASCADE"), nullable=False, index=True)
    department_id = Column(CHAR(36), ForeignKey("hospital_departments.id"), nullable=True)
    bed_number = Column(String(20), nullable=False)
    bed_type = Column(SQLEnum("general", "semi-private", "private", "icu", "nicu", "picu", "cardiac-icu", "emergency", "pediatric", "maternity"))
    floor = Column(String(50))
    ward = Column(String(100))
    status = Column(SQLEnum("available", "occupied", "reserved", "maintenance", "cleaning"), default="available")
    price_per_day = Column(DECIMAL(10, 2), default=0)
    features = Column(JSON)
    patient_id = Column(CHAR(36), ForeignKey("users.id"), nullable=True)
    patient_name = Column(String(150))
    admission_date = Column(Date)
    expected_discharge = Column(Date)
    has_ventilator = Column(Boolean, default=False)
    has_monitor = Column(Boolean, default=False)
    daily_charge = Column(DECIMAL(10, 2), default=0)


# ============================================
# HOSPITAL BLOOD BANK
# ============================================
class HospitalBloodBank(Base, BaseModel):
    __tablename__ = "hospital_blood_bank"

    hospital_id = Column(CHAR(36), ForeignKey("hospitals.id", ondelete="CASCADE"), unique=True, nullable=False, index=True)
    is_available = Column(Boolean, default=True)
    total_units = Column(Integer, default=0)
    expiry_alerts = Column(Integer, default=0)
    last_updated = Column(String(30))


# ============================================
# HOSPITAL BLOOD STOCKS
# ============================================
class HospitalBloodStock(Base, BaseModel):
    __tablename__ = "hospital_blood_stocks"

    blood_bank_id = Column(CHAR(36), ForeignKey("hospital_blood_bank.id", ondelete="CASCADE"), nullable=False, index=True)
    blood_group = Column(SQLEnum("A+", "A-", "B+", "B-", "AB+", "AB-", "O+", "O-"), nullable=False)
    units = Column(Integer, default=0)
    expiry_date = Column(Date)
    status = Column(SQLEnum("sufficient", "low", "critical", "out-of-stock"), default="sufficient")

    __table_args__ = (Index("ix_blood_stock_bank_group", "blood_bank_id", "blood_group"),)


# ============================================
# HOSPITAL OXYGEN STOCK
# ============================================
class HospitalOxygenStock(Base, BaseModel):
    __tablename__ = "hospital_oxygen_stock"

    hospital_id = Column(CHAR(36), ForeignKey("hospitals.id", ondelete="CASCADE"), unique=True, nullable=False, index=True)
    total_cylinders = Column(Integer, default=0)
    available_cylinders = Column(Integer, default=0)
    in_use_cylinders = Column(Integer, default=0)
    reserved_cylinders = Column(Integer, default=0)
    cylinder_types = Column(JSON)
    last_refilled = Column(Date)
    next_refill_date = Column(Date)
    supplier = Column(String(255))
    supplier_contact = Column(String(20))
    status = Column(SQLEnum("sufficient", "low", "critical", "out-of-stock"), default="sufficient")


# ============================================
# HOSPITAL AMBULANCES
# ============================================
class HospitalAmbulance(Base, BaseModel):
    __tablename__ = "hospital_ambulances"

    hospital_id = Column(CHAR(36), ForeignKey("hospitals.id", ondelete="CASCADE"), nullable=False, index=True)
    vehicle_number = Column(String(50), unique=True, nullable=False)
    type = Column(SQLEnum("basic", "advanced", "cardiac", "neonatal", "mobile-icu"), default="basic")
    status = Column(SQLEnum("available", "on-call", "dispatched", "maintenance", "offline"), default="available")
    driver_name = Column(String(150))
    driver_phone = Column(String(20))
    paramedic_name = Column(String(150))
    equipment = Column(JSON)
    current_latitude = Column(DECIMAL(10, 7))
    current_longitude = Column(DECIMAL(10, 7))
    current_address = Column(String(255))
    has_oxygen = Column(Boolean, default=True)
    has_ac = Column(Boolean, default=True)


# ============================================
# HOSPITAL ACTIVITIES
# ============================================
class HospitalActivity(Base, BaseModel):
    __tablename__ = "hospital_activities"

    hospital_id = Column(CHAR(36), ForeignKey("hospitals.id", ondelete="CASCADE"), nullable=False, index=True)
    type = Column(SQLEnum("admission", "discharge", "emergency", "surgery", "blood", "ambulance", "other"), nullable=False)
    description = Column(Text, nullable=False)
    department = Column(String(100))
    user_name = Column(String(150))
